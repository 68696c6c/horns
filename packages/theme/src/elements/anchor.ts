import * as Colorway from "../colorway"
import * as Outline from "../outline"
import * as Typography from "../typography"

import * as Decoration from "../tokens/decoration"
import * as Gap from "../tokens/gap"
import {
  Anchor,
  AnchorStateVariants,
  Ordinal,
  OrdinalVariants,
  ScaleVariants,
} from "../types"
import { Foundations } from "../foundations"

export type Variant = Ordinal

type VariantTokens = Colorway.Tokens & Decoration.Tokens

export type Element = Anchor<
  VariantTokens & Typography.Tokens,
  Outline.Tokens,
  VariantTokens,
  OrdinalVariants<AnchorStateVariants<Colorway.Tokens>> &
    ScaleVariants<Typography.ScaleTokens & Gap.Tokens>
>

type VariantConfigTokens = Colorway.ElementConfig & Decoration.ConfigTokens

export type Config = Anchor<
  VariantConfigTokens & Typography.ElementConfig,
  Outline.Variant,
  VariantConfigTokens,
  OrdinalVariants<AnchorStateVariants<Colorway.ElementConfig>> &
    ScaleVariants<Gap.ElementConfig>
>

export const makeElement = (f: Foundations, c: Config): Element => ({
  element: {
    ...Colorway.makeTokens(f.colorway, c.element),
    ...Typography.makeTokens(f.typography, c.element),
    ...Decoration.makeTokens(f.line, c.element),
  },
  focus: {
    ...Colorway.makeOutlineTokens(f.colorway),
    ...Outline.makeTokens(f.line, f.spacing, f.outline, c.focus),
  },
  hover: {
    ...Colorway.makeTokens(f.colorway, c.hover),
    ...Decoration.makeTokens(f.line, c.hover),
  },
  active: {
    ...Colorway.makeTokens(f.colorway, c.active),
    ...Decoration.makeTokens(f.line, c.active),
  },
  visited: {
    ...Colorway.makeTokens(f.colorway, c.visited),
    ...Decoration.makeTokens(f.line, c.visited),
  },
  current: {
    ...Colorway.makeTokens(f.colorway, c.visited),
    ...Decoration.makeTokens(f.line, c.visited),
  },
  variants: {
    primary: {
      base: Colorway.makeTokens(f.colorway, c.variants.primary.base),
      hover: Colorway.makeTokens(f.colorway, c.variants.primary.hover),
      active: Colorway.makeTokens(f.colorway, c.variants.primary.active),
      visited: Colorway.makeTokens(f.colorway, c.variants.primary.visited),
    },
    secondary: {
      base: Colorway.makeTokens(f.colorway, c.variants.secondary.base),
      hover: Colorway.makeTokens(f.colorway, c.variants.secondary.hover),
      active: Colorway.makeTokens(f.colorway, c.variants.secondary.active),
      visited: Colorway.makeTokens(f.colorway, c.variants.secondary.visited),
    },
    tertiary: {
      base: Colorway.makeTokens(f.colorway, c.variants.tertiary.base),
      hover: Colorway.makeTokens(f.colorway, c.variants.tertiary.hover),
      active: Colorway.makeTokens(f.colorway, c.variants.tertiary.active),
      visited: Colorway.makeTokens(f.colorway, c.variants.tertiary.visited),
    },
    sm: {
      ...Gap.makeTokens2(f.scale, c.variants.sm),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "sm"),
    },
    md: {
      ...Gap.makeTokens2(f.scale, c.variants.md),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "md"),
    },
    lg: {
      ...Gap.makeTokens2(f.scale, c.variants.lg),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "lg"),
    },
  },
})

export const defaultConfig: Config = {
  element: {
    colorway: "background.base",
    font: "body1",
    textDecorationLine: "underline",
    textDecorationStyle: "solid",
    textDecorationThickness: "02",
  },
  focus: "focus",
  hover: {
    colorway: "background.hover",
    textDecorationLine: "underline",
    textDecorationStyle: "double",
    textDecorationThickness: "02",
  },
  active: {
    colorway: "background.active",
    textDecorationLine: "underline",
    textDecorationStyle: "double",
    textDecorationThickness: "02",
  },
  visited: {
    colorway: "background.active",
    textDecorationLine: "underline",
    textDecorationStyle: "solid",
    textDecorationThickness: "02",
  },
  current: {
    colorway: "background.active",
    textDecorationLine: "underline",
    textDecorationStyle: "double",
    textDecorationThickness: "02",
  },
  variants: {
    primary: {
      base: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand1", "05"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand1", "04"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand1", "03"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      visited: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand1", "02"],
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    secondary: {
      base: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand2", "05"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand2", "04"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand2", "03"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      visited: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand2", "02"],
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    tertiary: {
      base: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand3", "05"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand3", "04"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand3", "03"],
        fill: "currentColor",
        stroke: "currentColor",
      },
      visited: {
        backgroundColor: "transparent",
        borderColor: "transparent",
        color: "currentColor",
        textDecorationColor: ["brand3", "02"],
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    sm: {
      // gap: "01",
      gap: "sm",
    },
    md: {
      // gap: "02",
      gap: "md",
    },
    lg: {
      // gap: "03",
      gap: "lg",
    },
  },
}
