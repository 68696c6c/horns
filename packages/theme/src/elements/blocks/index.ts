import {
  BorderStyle,
  Breakpoint,
  Color,
  evalCornerSizesConfigs,
  evalShadows,
  evalSideBordersConfigs,
  evalSideSizesConfigs,
  Font,
  ShadowType,
  Size,
} from '../../config'
import { Cursor, DeepPartial, evalConfigBooleans } from '../../utils'

import * as Element from '../element'

type Block<BaseType> = BaseType & {
  breakpoint: Breakpoint
  contained: boolean
}

export interface Config extends Block<Element.Config> {}

export interface Props extends Block<Element.Props> {}

export interface Theme extends Block<Element.Theme> {}

export const defaultConfig: Config = {
  breakpoint: Breakpoint.Mobile,
  border: {
    all: {
      width: Size.None,
      style: BorderStyle.None,
    },
  },
  color: Color.BgPrimary,
  contained: false,
  cursor: Cursor.Default,
  font: Font.Text,
  padding: {
    all: Size.XSmall,
  },
  radius: {
    all: Size.None,
  },
  shadowed: false,
  shadowType: ShadowType.Box,
}

export const make = (input?: DeepPartial<Config>): Theme => ({
  breakpoint: input?.breakpoint || defaultConfig.breakpoint,
  color: input?.color || defaultConfig.color,
  contained: evalConfigBooleans(defaultConfig.contained, 'contained', input),
  cursor: input?.cursor || defaultConfig.cursor,
  border: evalSideBordersConfigs(defaultConfig.border, input?.border),
  font: input?.font || defaultConfig.font,
  padding: evalSideSizesConfigs(defaultConfig.padding, input?.padding),
  radius: evalCornerSizesConfigs(defaultConfig.radius, input?.radius),
  shadow: evalShadows(defaultConfig.shadowType, defaultConfig.shadowed, input),
})
