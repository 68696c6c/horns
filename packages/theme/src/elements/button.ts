import * as Border from "../border"
import * as Colorway from "../colorway"
import * as Outline from "../outline"
import * as Radius from "../radius"
import * as Typography from "../typography"

import * as Gap from "../tokens/gap"
import * as Padding from "../tokens/padding"
import { Control, ControlStateVariants, Scale, ScaleVariants } from "../types"
import { Foundations } from "../foundations"
import { ElementConfig, makeTokens2 } from "../tokens/padding"

type VariantTokens = Border.Tokens & Colorway.Tokens & Radius.Tokens

export type Element = Control<
  VariantTokens & Typography.Tokens,
  Outline.Tokens,
  VariantTokens,
  ButtonVariants & ButtonScaleVariants
>

type VariantConfigTokens = Partial<Border.ElementConfig> &
  Colorway.ElementConfig &
  Partial<Radius.ElementConfig>

export type Config = Control<
  Colorway.ElementConfig &
    Border.ElementConfig &
    Radius.ElementConfig &
    Typography.ElementConfig,
  Outline.Variant,
  VariantConfigTokens,
  ButtonVariants<VariantConfigTokens, Colorway.ElementConfig> &
    ButtonScaleVariants<
      Gap.ElementConfig & Padding.ElementConfig,
      Gap.ElementConfig & Padding.ElementConfig
    >
>

type ButtonVariants<S = VariantTokens, F = Colorway.Tokens> = {
  primary: ControlStateVariants<S>
  secondary: ControlStateVariants<S>
  tertiary: ControlStateVariants<S>
  enabled: ControlStateVariants<S>
  disabled: F
  error: F
}

type ButtonScaleVariants<
  M = Gap.Tokens & Padding.Tokens,
  S = Typography.ScaleTokens & Gap.Tokens & Padding.Tokens
> = ScaleVariants<S> & {
  min: M
}

export type Variant = keyof ButtonVariants

// export type ScaleVariant = "min" | Scale

const makeVariantTokens = (
  f: Foundations,
  c: VariantConfigTokens
): VariantTokens => ({
  ...Border.makeTokens(f.line, f.border, c),
  ...Colorway.makeTokens(f.colorway, c),
  ...Radius.makeTokens(f.radius, c),
})

export const makeElement = (f: Foundations, c: Config): Element => ({
  element: {
    ...makeVariantTokens(f, c.element),
    ...Typography.makeTokens(f.typography, c.element),
  },
  focus: {
    ...Colorway.makeOutlineTokens(f.colorway),
    ...Outline.makeTokens(f.line, f.spacing, f.outline, c.focus),
  },
  hover: makeVariantTokens(f, c.hover),
  active: makeVariantTokens(f, c.active),
  variants: {
    primary: {
      base: makeVariantTokens(f, c.variants.primary.base),
      hover: makeVariantTokens(f, c.variants.primary.hover),
      active: makeVariantTokens(f, c.variants.primary.active),
    },
    secondary: {
      base: makeVariantTokens(f, c.variants.secondary.base),
      hover: makeVariantTokens(f, c.variants.secondary.hover),
      active: makeVariantTokens(f, c.variants.secondary.active),
    },
    tertiary: {
      base: makeVariantTokens(f, c.variants.tertiary.base),
      hover: makeVariantTokens(f, c.variants.tertiary.hover),
      active: makeVariantTokens(f, c.variants.tertiary.active),
    },
    enabled: {
      base: makeVariantTokens(f, c.variants.enabled.base),
      hover: makeVariantTokens(f, c.variants.enabled.hover),
      active: makeVariantTokens(f, c.variants.enabled.active),
    },
    disabled: makeVariantTokens(f, c.variants.disabled),
    error: makeVariantTokens(f, c.variants.error),
    min: {
      // ...Gap.makeTokens(f.spacing, c.variants.min),
      ...Gap.makeTokens2(f.scale, c.variants.min),
      // ...Padding.makeTokens(f.spacing, c.variants.min),
      ...Padding.makeTokens2(f.scale, c.variants.min),
    },
    sm: {
      // ...Gap.makeTokens(f.spacing, c.variants.sm),
      ...Gap.makeTokens2(f.scale, c.variants.sm),
      // ...Padding.makeTokens(f.spacing, c.variants.sm),
      ...Padding.makeTokens2(f.scale, c.variants.sm),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "sm"),
    },
    md: {
      // ...Gap.makeTokens(f.spacing, c.variants.md),
      ...Gap.makeTokens2(f.scale, c.variants.md),
      // ...Padding.makeTokens(f.spacing, c.variants.md),
      ...Padding.makeTokens2(f.scale, c.variants.md),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "md"),
    },
    lg: {
      // ...Gap.makeTokens(f.spacing, c.variants.lg),
      ...Gap.makeTokens2(f.scale, c.variants.lg),
      // ...Padding.makeTokens(f.spacing, c.variants.lg),
      ...Padding.makeTokens2(f.scale, c.variants.lg),
      ...Typography.makeScaleTokens(f.typography, c.element.font, "lg"),
    },
  },
})

export const defaultConfig: Config = {
  element: {
    border: "none",
    colorway: "background.base",
    font: "body1",
    radius: "normal",
  },
  focus: "focus",
  hover: {
    colorway: "background.hover",
  },
  active: {
    colorway: "background.active",
  },
  variants: {
    primary: {
      base: {
        border: "medium",
        backgroundColor: ["brand1", "05"],
        borderColor: ["brand1", "04"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        backgroundColor: ["brand1", "04"],
        borderColor: ["brand1", "03"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        backgroundColor: ["brand1", "03"],
        borderColor: ["brand1", "02"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    secondary: {
      base: {
        border: "thick",
        backgroundColor: "transparent",
        borderColor: ["brand2", "05"],
        color: ["brand2", "05"],
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        // border: "thick",
        backgroundColor: "transparent",
        borderColor: ["brand2", "04"],
        color: ["brand2", "04"],
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        // border: "thick",
        backgroundColor: "transparent",
        borderColor: ["brand2", "03"],
        color: ["brand2", "03"],
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    tertiary: {
      base: {
        border: "thin",
        backgroundColor: ["brand3", "05"],
        borderColor: ["brand3", "04"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      hover: {
        backgroundColor: ["brand3", "04"],
        borderColor: ["brand3", "03"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
      active: {
        backgroundColor: ["brand3", "03"],
        borderColor: ["brand3", "02"],
        color: "readable",
        textDecorationColor: "currentColor",
        fill: "currentColor",
        stroke: "currentColor",
      },
    },
    enabled: {
      base: {
        colorway: "enabled.base",
      },
      hover: {
        colorway: "enabled.hover",
      },
      active: {
        colorway: "enabled.active",
      },
    },
    disabled: {
      colorway: "disabled.base",
    },
    error: {
      colorway: "error.base",
    },
    min: {
      // gap: "00",
      gap: "min",
      padding: "min",
      // paddingTop: "00",
      // paddingBottom: "00",
      // paddingLeft: "00",
      // paddingRight: "00",
    },
    sm: {
      // gap: "01",
      gap: "sm",
      padding: "sm",
      // paddingTop: "01",
      // paddingBottom: "01",
      // paddingLeft: "02",
      // paddingRight: "02",
    },
    md: {
      // gap: "02",
      gap: "md",
      padding: "md",
      // paddingTop: "02",
      // paddingBottom: "02",
      // paddingLeft: "03",
      // paddingRight: "03",
    },
    lg: {
      // gap: "03",
      gap: "lg",
      padding: "lg",
      // paddingTop: "03",
      // paddingBottom: "03",
      // paddingLeft: "04",
      // paddingRight: "04",
    },
  },
}
