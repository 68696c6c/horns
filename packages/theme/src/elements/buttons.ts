import { BorderStyle, evalSideBordersConfigs } from '../borders'
import { Color } from '../colors'
import { ShadowType } from '../shadows'
import { Size, evalCornerSizesConfigs, evalSideSizesConfigs } from '../sizes'
import { Font } from '../typography'
import { Cursor } from '../utils'

import {
  ElementConfig,
  ElementProps,
  ThemeElement,
  evalBooleanProp,
} from './elements'

export const defaultButtons: ElementConfig = {
  border: {
    all: {
      width: Size.Tiny,
      style: BorderStyle.Solid,
    },
  },
  color: Color.BgPrimary,
  cursor: Cursor.Pointer,
  font: Font.Control,
  outlined: false,
  padding: {
    x: Size.Medium,
    y: Size.Small,
  },
  radius: {
    all: Size.Tiny,
  },
  shadowed: false,
  shadowType: ShadowType.Box,
  typographic: false,
}

export const makeButtons = (input?: ElementProps): ThemeElement => ({
  color: input?.color || defaultButtons.color,
  cursor: input?.cursor || defaultButtons.cursor,
  border: evalSideBordersConfigs(defaultButtons.border, input?.border),
  font: input?.font || defaultButtons.font,
  outlined: evalBooleanProp(defaultButtons.outlined, 'outlined', input),
  padding: evalSideSizesConfigs(defaultButtons.padding, input?.padding),
  shadow: input?.shadowed ? defaultButtons.shadowType : ShadowType.None,
  radius: evalCornerSizesConfigs(defaultButtons.radius, input?.radius),
  typographic: defaultButtons.typographic,
})
