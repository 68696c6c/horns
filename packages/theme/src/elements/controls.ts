import { BorderStyle, evalSideBordersConfigs } from '../borders'
import { Color } from '../colors'
import { ShadowType } from '../shadows'
import { Size, evalCornerSizesConfigs, evalSideSizesConfigs } from '../sizes'
import { Font } from '../typography'
import { Cursor } from '../utils'

import {
  ElementConfig,
  ElementProps,
  ThemeElement,
  evalBooleanProp,
} from './elements'

export const defaultControls: ElementConfig = {
  border: {
    all: {
      width: Size.Tiny,
      style: BorderStyle.Solid,
    },
  },
  color: Color.BgPrimary,
  cursor: Cursor.Default,
  font: Font.Control,
  outlined: false,
  padding: {
    all: Size.XSmall,
  },
  radius: {
    all: Size.Tiny,
  },
  shadowed: false,
  shadowType: ShadowType.Inset,
  typographic: false,
}

export const makeControls = (input?: ElementProps): ThemeElement => ({
  color: input?.color || defaultControls.color,
  cursor: input?.cursor || defaultControls.cursor,
  border: evalSideBordersConfigs(defaultControls.border, input?.border),
  font: input?.font || defaultControls.font,
  outlined: evalBooleanProp(defaultControls.outlined, 'outlined', input),
  padding: evalSideSizesConfigs(defaultControls.padding, input?.padding),
  shadow: input?.shadowed ? defaultControls.shadowType : ShadowType.None,
  radius: evalCornerSizesConfigs(defaultControls.radius, input?.radius),
  typographic: defaultControls.typographic,
})
