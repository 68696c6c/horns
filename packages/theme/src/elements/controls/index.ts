import {
  BorderStyle,
  Color,
  ControlColor,
  evalCornerSizesConfigs,
  evalShadows,
  evalSideBordersConfigs,
  evalSideSizesConfigs,
  Font,
  ShadowType,
  Size,
} from '../../config'
import { Cursor, DeepPartial, evalConfigBooleans } from '../../utils'

import * as Interactive from '../interactive'

type Control<
  BaseType extends Interactive.Config | Interactive.Props | Interactive.Theme,
> = Omit<BaseType, 'color'> & {
  color: ControlColor
}

export interface Config extends Control<Interactive.Config> {}

export interface Props extends Control<Interactive.Props> {}

export interface Theme extends Control<Interactive.Theme> {}

export const defaultConfig: Config = {
  border: {
    all: {
      width: Size.Tiny,
      style: BorderStyle.Solid,
    },
  },
  color: Color.BgPrimary,
  cursor: Cursor.Default,
  font: Font.Control,
  outlined: false,
  padding: {
    all: Size.XSmall,
  },
  radius: {
    all: Size.Tiny,
  },
  shadowed: false,
  shadowType: ShadowType.Inset,
  typographic: false,
}

export const make = (input?: DeepPartial<Config>): Theme => ({
  border: evalSideBordersConfigs(defaultConfig.border, input?.border),
  color: input?.color || defaultConfig.color,
  cursor: input?.cursor || defaultConfig.cursor,
  font: input?.font || defaultConfig.font,
  outlined: evalConfigBooleans(defaultConfig.outlined, 'outlined', input),
  padding: evalSideSizesConfigs(defaultConfig.padding, input?.padding),
  radius: evalCornerSizesConfigs(defaultConfig.radius, input?.radius),
  shadow: evalShadows(defaultConfig.shadowType, defaultConfig.shadowed, input),
  typographic: defaultConfig.typographic,
})
