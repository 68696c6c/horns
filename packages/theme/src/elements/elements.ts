import { SideBorders, SideBordersConfig } from '../borders'
import { Color } from '../colors'
import { ShadowType } from '../shadows'
import {
  CornerSizes,
  CornerSizesConfig,
  SideSizes,
  SideSizesConfig,
} from '../sizes'
import { Font } from '../typography'
import { Cursor, DeepPartial } from '../utils'

export interface ElementConfig {
  color: Color
  cursor: Cursor
  border: SideBordersConfig
  font: Font
  padding: SideSizesConfig
  radius: CornerSizesConfig
  shadowed: boolean
  shadowType: ShadowType
  // Typographic elements use the "color" prop to set the color of the text; no background color is
  // set.  Since the system is unaware of the background color, it cannot ensure the text will be
  // readable.  In typographic elements, the "outlined" prop enables a text outline that helps
  // ensure the readability of the text regardless of background color.
  //
  // Non-typographic elements instead use the "color" prop to set the background color of the
  // element and the text color is either black or white depending on which is most readable. In
  // non-typographic components, the "outlined" prop causes the "color" prop to control the border
  // and text color instead of the background.
  typographic: boolean
  outlined: boolean
}

// This type is intended to be used for component props and should therefore represent everything
// the end user is allowed to configure or control about an element.
//
// Typographic is omitted because the system needs to be able to reliably make certain assumptions
// about text color behavior at the element level and trying to support user input in that process
// creates potential for unpredictable behavior.  See ElementConfig comments.
//
// The "shadowType" prop is omitted because components use the "shadowed" prop to enable a theme-
// level "shadowType" set at the element level.  All instances of an element use the same shadow.
export type ElementProps = DeepPartial<
  Omit<ElementConfig, 'typographic' | 'shadowType'>
>

export interface ThemeElement {
  color: Color
  cursor: Cursor
  border: SideBorders
  font: Font
  outlined: boolean
  padding: SideSizes
  shadow: ShadowType
  radius: CornerSizes
  typographic: boolean
}

export const evalBooleanProp = (
  defaultValue: boolean,
  key: keyof Pick<ElementProps, 'outlined' | 'shadowed'>,
  config?: ElementProps,
): boolean =>
  config && typeof config[key] === 'boolean'
    ? (config[key] as boolean)
    : defaultValue
