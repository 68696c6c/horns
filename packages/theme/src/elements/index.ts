export { ElementConfig, ElementProps, ThemeElement } from './elements'
export { makeButtons } from './buttons'
export { makeControls } from './controls'
export { makeLinks } from './links'
export { NavConfig, NavProps, NavTheme, makeNav } from './nav'
export { makeTables } from './tables'
