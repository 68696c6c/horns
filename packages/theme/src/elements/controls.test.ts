import { BorderStyle } from '../borders'
import { Color } from '../colors'
import { Cursor } from '../cursors'
import { Size } from '../sizes'
import { Font } from '../typography'

import { defaultControls, makeControls } from './controls'
import { ElementConfig } from './elements'

describe('makeControls', () => {
  const expected = makeControls(defaultControls)

  it('should return default values if no input is provided', () => {
    const result = makeControls()
    expect(result).toStrictEqual(expected)
  })

  it('should use default values for incomplete inputs', () => {
    const input: Partial<ElementConfig> = {
      color: Color.Tertiary,
    }
    const result = makeControls(input)
    expect(result).toMatchSnapshot()
  })

  describe('border', () => {
    it('should accept "all" inputs', () => {
      const inputBorder = {
        style: BorderStyle.Groove,
        width: Size.Giant,
      }
      const input: Partial<ElementConfig> = {
        border: {
          all: inputBorder,
        },
      }
      const result = makeControls(input)
      expect(result.border).toStrictEqual({
        top: inputBorder,
        bottom: inputBorder,
        left: inputBorder,
        right: inputBorder,
      })
    })

    it('should accept dimensional inputs', () => {
      const input: Partial<ElementConfig> = {
        border: {
          x: {
            style: BorderStyle.Groove,
            width: Size.Giant,
          },
          y: {
            style: BorderStyle.Double,
            width: Size.XXSmall,
          },
        },
      }
      const result = makeControls(input)
      expect(result.border).toStrictEqual({
        top: {
          style: BorderStyle.Double,
          width: Size.XXSmall,
        },
        bottom: {
          style: BorderStyle.Double,
          width: Size.XXSmall,
        },
        left: {
          style: BorderStyle.Groove,
          width: Size.Giant,
        },
        right: {
          style: BorderStyle.Groove,
          width: Size.Giant,
        },
      })
    })

    it('should accept complete inputs', () => {
      const input: Partial<ElementConfig> = {
        border: {
          top: {
            style: BorderStyle.Double,
            width: Size.Giant,
          },
          bottom: {
            style: BorderStyle.Dashed,
            width: Size.XXLarge,
          },
          left: {
            style: BorderStyle.Groove,
            width: Size.XLarge,
          },
          right: {
            style: BorderStyle.Dotted,
            width: Size.Medium,
          },
        },
      }
      const result = makeControls(input)
      expect(result.border).toStrictEqual({
        top: {
          style: BorderStyle.Double,
          width: Size.Giant,
        },
        bottom: {
          style: BorderStyle.Dashed,
          width: Size.XXLarge,
        },
        left: {
          style: BorderStyle.Groove,
          width: Size.XLarge,
        },
        right: {
          style: BorderStyle.Dotted,
          width: Size.Medium,
        },
      })
    })

    it('should merge nav and item inputs', () => {
      const input: Partial<ElementConfig> = {
        border: {
          all: {
            style: BorderStyle.Double,
            width: Size.Giant,
          },
        },
      }
      const result = makeControls(input)
      expect(result.border).toStrictEqual({
        top: {
          style: BorderStyle.Double,
          width: Size.Giant,
        },
        bottom: {
          style: BorderStyle.Double,
          width: Size.Giant,
        },
        left: {
          style: BorderStyle.Double,
          width: Size.Giant,
        },
        right: {
          style: BorderStyle.Double,
          width: Size.Giant,
        },
      })
    })
  })

  it('should accept color values', () => {
    const input: Partial<ElementConfig> = {
      color: Color.Tertiary,
    }
    const result = makeControls(input)
    expect(result.color).toBe(Color.Tertiary)
  })

  it('should accept cursor values', () => {
    const input: Partial<ElementConfig> = {
      cursor: Cursor.Pointer,
    }
    const result = makeControls(input)
    expect(result.cursor).toBe(Cursor.Pointer)
  })

  it('should accept font values', () => {
    const input: Partial<ElementConfig> = {
      font: Font.Control,
    }
    const result = makeControls(input)
    expect(result.font).toBe(Font.Control)
  })

  describe('padding', () => {
    it('should accept "all" inputs', () => {
      const input: Partial<ElementConfig> = {
        padding: { all: Size.Giant },
      }
      const result = makeControls(input)
      expect(result.padding).toStrictEqual({
        top: Size.Giant,
        bottom: Size.Giant,
        left: Size.Giant,
        right: Size.Giant,
      })
    })

    it('should accept dimensional inputs', () => {
      const input: Partial<ElementConfig> = {
        padding: {
          x: Size.Giant,
          y: Size.Large,
        },
      }
      const result = makeControls(input)
      expect(result.padding).toStrictEqual({
        top: Size.Large,
        bottom: Size.Large,
        left: Size.Giant,
        right: Size.Giant,
      })
    })

    it('should accept complete inputs', () => {
      const input: Partial<ElementConfig> = {
        padding: {
          top: Size.Giant,
          bottom: Size.Large,
          left: Size.Medium,
          right: Size.Small,
        },
      }
      const result = makeControls(input)
      expect(result.padding).toStrictEqual({
        top: Size.Giant,
        bottom: Size.Large,
        left: Size.Medium,
        right: Size.Small,
      })
    })

    it('should merge nav and item inputs', () => {
      const input: Partial<ElementConfig> = {
        padding: { all: Size.Giant },
      }
      const result = makeControls(input)
      expect(result.padding).toStrictEqual({
        top: Size.Giant,
        bottom: Size.Giant,
        left: Size.Giant,
        right: Size.Giant,
      })
    })
  })

  describe('radius', () => {
    it('should accept "all" inputs', () => {
      const input: Partial<ElementConfig> = {
        radius: { all: Size.Giant },
      }
      const result = makeControls(input)
      expect(result.radius).toStrictEqual({
        topRight: Size.Giant,
        bottomRight: Size.Giant,
        topLeft: Size.Giant,
        bottomLeft: Size.Giant,
      })
    })

    it('should accept dimensional inputs', () => {
      const input: Partial<ElementConfig> = {
        radius: {
          left: Size.Giant,
          right: Size.Large,
        },
      }
      const result = makeControls(input)
      expect(result.radius).toStrictEqual({
        topRight: Size.Large,
        bottomRight: Size.Large,
        topLeft: Size.Giant,
        bottomLeft: Size.Giant,
      })
    })

    it('should accept complete inputs', () => {
      const input: Partial<ElementConfig> = {
        radius: {
          topRight: Size.Giant,
          bottomRight: Size.Large,
          topLeft: Size.Medium,
          bottomLeft: Size.Small,
        },
      }
      const result = makeControls(input)
      expect(result.radius).toStrictEqual({
        topRight: Size.Giant,
        bottomRight: Size.Large,
        topLeft: Size.Medium,
        bottomLeft: Size.Small,
      })
    })

    it('should merge nav and item inputs', () => {
      const input: Partial<ElementConfig> = {
        radius: { all: Size.Giant },
      }
      const result = makeControls(input)
      expect(result.radius).toStrictEqual({
        topRight: Size.Giant,
        bottomRight: Size.Giant,
        topLeft: Size.Giant,
        bottomLeft: Size.Giant,
      })
    })
  })
})
