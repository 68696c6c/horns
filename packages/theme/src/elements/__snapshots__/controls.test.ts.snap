// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeControls should use default values for incomplete inputs 1`] = `
Object {
  "border": Object {
    "bottom": Object {
      "style": "solid",
      "width": "tiny",
    },
    "left": Object {
      "style": "solid",
      "width": "tiny",
    },
    "right": Object {
      "style": "solid",
      "width": "tiny",
    },
    "top": Object {
      "style": "solid",
      "width": "tiny",
    },
  },
  "color": "tertiary",
  "cursor": "default",
  "font": "control",
  "outlined": false,
  "padding": Object {
    "bottom": "xSmall",
    "left": "xSmall",
    "right": "xSmall",
    "top": "xSmall",
  },
  "radius": Object {
    "bottomLeft": "tiny",
    "bottomRight": "tiny",
    "topLeft": "tiny",
    "topRight": "tiny",
  },
  "shadow": "none",
  "typographic": false,
}
`;
